### PR TITLE
rename several network entities

### PR DESCRIPTION
**Pull request by dankenigsberg**
This is a follow up of PR #4775. Much like its predecessor, it includes multiple commits, each introducing a name change or code deletion. The most notable changes here are
* Various implementations of BindMechanism are named accordingly, so it is easier to discern their interface
* The aggregate of the two PlugPhase functions is renamed to "PlugPhases" since its current name (NetworkInterface) is too overloaded

I hope these changes make the code more accessible to newcomers like myself.

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/network.go]: IMO this factory has no place at all, it does nothing in practice and returning an interface is just wrong.
But if we do not want to go that far, then I would just recommend to drop the `get` prefix. `newNetworkInterface` sounds enough.

File: pkg/virt-launcher/virtwrap/network/podinterface.go
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/podinterface.go:370]: I think the  whole binding mechanism should be exported to its own package (e.g. `bindmechanism` or `bindmethod`).
And then each concrete struct and method can be implemented in a different file.

Then usage will be: `bind := bindmechanism.Bridge{...}`

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/network.go:99]: The name hints its usage (to be mocked) and not what it does.
Having in the production code a function name "mockable" is too odd IMO.

Converting it to a private method sounds more appropriate to me.
On the other hand, removing the globals should be the primary goal, but I do recognize it requires much more work.

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/network.go:64]: I think an interface should be a noun in its single form.
Making it plural hints it contains something.

If this interface represents something (I find it hard to see something useful really), then it can be `NetworkConfig` or `PodNetworkConfig`, whatever that means.

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/network.go:136]: No need for the `get` prefix here.

**Review by EdDev — CHANGES_REQUESTED** [review]
I have mixed feelings about trying to name better something that makes little sense in terms of structure.
Some are even renamed twice in the same PR (in different commits).

But I guess it cannot make it worse. If it makes it even a bit more readable, then lets go for it.

Commit notes:
- `network: drop NetworkInterface.Unplug` :
The sentence needs fixing: `In the current moment
it time it only...`

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go]: I share your opinion about the future fate of this "factory" that produces one instance of a struct with no fields. This PR is supposed to be a minor step forward, calling entities for what they are. Future PRs should actually change these entities and how they relate to each other.

File: pkg/virt-launcher/virtwrap/network/podinterface.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/podinterface.go:370]: this makes sense to me, but I prefer to do this in a follow-up PR (certainly on a follow-up commit)

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go:99]: The `mockable` function name is very odd, but the problem is not with its name. The problem is that it exists only because we want to mock it in the tests. This PR wants to put these oddities on the surface, so every reader of the code immediately understands why this function exist. For me, it took few hours and few phonecalls to understand that this function exists in production only because the test wants it. I would like that future readers would not have this ordeal again.

The name keeps a hit about what it does `SetupPodNetworkPhase1`, though one would argue that `Phase1` is an opaque term. This PR does not try to fix everything, only few of them, one at a time.

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: I find it very hard to fathom what this interface represents. I tried to make it very clear in the commit message
```
    An instance of network.NetworkInterface does not represent an actual
    network interface; it is just an aggregation of two methods, used to
    plug a network interface to a pod. However in a few cases, the name
    `vif` is unjustly attached to such instance.
```
I don't think that it is a NetworkConfig. It is just a pare of loosely-related functions exposed so they easier to use by gomock. I is hard to find a non-awkward name to an entity with such an awkward raison d'etre. How about `PlugPhasePair`?

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:99]: You will indeed expose the mess, but on the other hand I am unsure if new ones who read the code (even me), will really understand why we have a `mockable` function.
I suggest to add this info through a comment.

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: I was thinking that one can saw: I pass you a NetworkConfig, on which you can plug (execute) two phases to apply it.
Maybe `NetworkConfigurer` ? The configurer has two plug-gable phases it can execute to apply it.

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go:99]: You forced me to look deeper into the code, and I find no reason to have this mocking at all. I'll push a branch which drops it.

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: It seems to me that you are trying to force logic into where it does not exist. I spent hours trying to understand how this `NetworkInterface` relates to similarly-named entities in the project. I failed. I don't want to mislead future readers to assume that there exists an entity named `NetworkConfig` that can be passed from a producer to a consumer and has any kind of standing. This `PlugPhases` is an aggregate of two functions, with no meaning beyond the possibility to mock them.

I hope that in a future PR, someone smarter than me can rewrite the tests so they do not require this clog.

Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: Ok, I give up.

**Review by EdDev — APPROVED** [review]
Thank you for the cleanup!

File: pkg/virt-launcher/virtwrap/network/common.go
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/network/common.go]: I think the same can be done for `SetupPodNetworkPhase2`.
Although there is a test actually modifying it, the test subject doesn't use `SetupPodNetworkPhase2` it is directly invoking `SetupNetworkInterfacesPhase2`.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Why aren't you also renaming the `macvtapPodInterface` as well in https://github.com/kubevirt/kubevirt/pull/4777/commits/d29e020abbf35e67bb0961c96717a104cfb19bb6 ? 

I'd say either you rename all, or none. Is there a reason to single out that (poor) bind mechanism ?

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: > I think an interface should be a noun in its single form.

I agree with Eddy.

@dankenigsberg @EdDev what do you think about -
1. change the name of `NetworkInterface` to `PodNic`.
2. change the name of `PodInterface` to `PodNicImpl`
3. Move the parameters passed to `PlugPhase1\2` to be data members of `PodInterfaceNicImpl`. So that `PodInterfaceNicImpl` won't be just an aggregation of two methods but actually will have a state.
4. In a follow-up PR/commit `{Bridge,Masquerade,Slirp}PodInterface` will contain `PodInterface` as a data member.
5. In a follow-up PR I would also change `PlugPhase1\2` to -
`PlugPhase1`-> `ConfigurePodInterfaceBindingViaVirtHandler`
`PlugPhase2`-> `ConfigurePodInterfaceBindingViaVirtLauncher`

Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: I think you are suggestion now: `PodNic.PlugPhase1`.
The problem is that it does not represent really a nic, the action is on all nics/interfaces of the VMI. (that is why I tried to suggest something generic like network-config.

At the end, I would like to see the phases just gone.
I do not think anyone should group the actions on the handler or on the launcher... the "grouping" is just the fact that they are called in the correct place/flow.

I think all suggestions are good if they move us one bit forward, so I would agree with anything at this point :)

Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: `PodNic.PlugPhase1` is on a specific nic. It is not on all nics/interfaces.

Anyway, after re-thinking, point 4 is problematic as `PodNic.PlugPhase1` is creating the binding so it cannot pass itself as a parameter to the binding. So ignore point 4. The other are still relevant.

Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: Thank you for clarifying and correcting me here. Passing in the VMI confused me.
As such, I like this option. The name `PodNIC` could also be `PodInterface` but it is a nit and both will work out for me.

Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: The reason I chose `Nic` and not `Interface` is to avoid confusion with the golang interface.

**Comment by dankenigsberg** [comment]
> Why aren't you also renaming the `macvtapPodInterface` as well in [d29e020](https://github.com/kubevirt/kubevirt/commit/d29e020abbf35e67bb0961c96717a104cfb19bb6) ?
> 
> I'd say either you rename all, or none. Is there a reason to single out that (poor) bind mechanism ?

All I can see is "oops".

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go:99]: The idea of phase1 and phase2 runs deeper than this, in the code and in comments. I sense that eliminating them may be a more contentious matter. I see, for example, that @AlonaKaplan is suggesting another name. I would prefer to do it in a separate commit: I think that linking `NetworkInterfaces` and `PodNetwork` together is important on its own right, and should not hold until phases are gone.

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go]: This PR starts with (hopefully) less contentious commits. I consider `getNetworkClass -> getNewNetworkInterface` as a very clear and obvious improvement. The next change `getNewNetworkInterface -> newPlugPhases` has very different (and deeper) reasoning, so I prefer to put it in a commit which you consider(ed?) unnecessary.

**Comment by dankenigsberg** [comment]
(I don't understand why some github comments may be replied to in-place, some have "quote", and this one has neither.)

> I think we can all agree this is a mess.
> Now, I don't think renaming this to an equally wrong term achieves anything on itself - other than possibly highlighting that this is wrong.
> IMHO, this commit is unnecessary .

I agree that both new and old names are bad, but they are NOT equally wrong. The current name makes people (like myself) to assume that there's purpose in the object passed, and that it is somehow related to other NetworkInterfaces in the code. It does not. The name I suggest makes it clear that we have an abstraction problem, which has to be fixed soon (preferably by someone smarter than me).

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go]: My problem with `generateNetworkInterface` is that it is hard to differentiate it from `getNewNetworkInterface`. There's a good reason for this confusing: there is no real justification for `NetworkInterfaceFactory`. But I try not to venture to these corners of the code, and rather keep this PR about renaming (and few code removals).

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: > 3. Move the parameters passed to PlugPhase1\2 to be data members of PodInterfaceNicImpl. So that PodInterfaceNicImpl won't be just an aggregation of two methods but actually will have a state.

To recap, do you suggest
```
PodNICImpl struct {
 vmi *v1.VirtualMachineInstance
 iface *v1.Interface
 network *v1.Network
 podInterfaceName string
}
``` 
(`pid` does not belong there, I assume) ?

The factory would have to be changed to fill these (somewhat redundant) fields. I feel that it is going to be a very long crawl in a deep rat hole until an object worthy of the `PodNIC` name can be created, which I'm not yet sure if should have one at all. I certainly cannot do this in this rename-only PR.

The ugly name `PlugPhases` is meant as a stopgap, a mark saying "this is not an object. this is bad. do not assume anything about this interface because nothing is there". Maybe `PodNIC` is obscure enough not to cause assumptions. I'll try.

Comment (dankenigsberg) [reply pkg/virt-launcher/virtwrap/network/network.go:64]: joining Phase1/2 renames suggested by @AlonaKaplan and @maiqueb for future consideration

> PlugPhase1-> ConfigurePodInterfaceBindingViaVirtHandler SetupPodNetworkTrustedComponentPhase or SetupPodNetworkPrivilegedPhase
> PlugPhase2-> ConfigurePodInterfaceBindingViaVirtLauncher SetupPodNetworkUntrustedComponentPhase or SetupPodNetworkUnprivilegedPhase

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/network/network.go:82]: Even though `podNic` is private, in the method name it is not relevant. So I would use - `invokePodNICFactory`.

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:99]: +1 for eliminating the phases completely in a future PR.
Just by calling the relevant functions in the virt-handler implies they are priv, there is no need to group them under this term.

**Review by EdDev — APPROVED** [review]

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (maiqueb) [reply pkg/virt-launcher/virtwrap/network/network.go:99]: > The idea of phase1 and phase2 runs deeper than this, in the code and in comments. I sense that eliminating them may be a more contentious matter. I see, for example, that @AlonaKaplan is suggesting another name.

Yes, a name that focused on 'who' rather than on 'why' . If the goal of this PR is to help whoever reads the code to better understand what this does, I'd say the 'why' is important.

> I would prefer to do it in a separate commit: I think that linking `NetworkInterfaces` and `PodNetwork` together is important on its own right, and should not hold until phases are gone.

> +1 for eliminating the phases completely in a future PR.

I'm OK with that.

> Just by calling the relevant functions in the virt-handler implies they are priv, there is no need to group them under this term.

Yes, but that is not happening here - by renaming these functions accordingly, you can convey information to the reader of why this is like it is. That adds value to this intermediate state, until this is eventually refactored. 
And is so cheap to do it (rename from A to B costs the same independently of what B is) that I don't understand why we wouldn't want to do it.

**Review by maiqueb — COMMENTED** [review]
/hold

Just read my comment, you can remove the hold if you think I'm not making a good point (but I'd like to understand why).

File: pkg/virt-launcher/virtwrap/network/network.go
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/network.go:99]: @maiqueb , it is impossible to explain it because it makes no sense.
I agree this change is not enough to make things better, but it is also not going into the wrong direction.
The name change was focusing on not confusing it with other things, which it partially did. Is to well describing what is going on? Sure it does not, no one can describe something that is doing a lot of things.

I think this change is not making things worse, it makes it a little bit less confusing without refactoring the core problem.

> Yes, but that is not happening here - by renaming these functions accordingly, you can convey information to the reader of why this is like it is. That adds value to this intermediate state, until this is eventually refactored.

I think the author is not interested to try and explain it, but rather try to avoid colliding the name with other similar functions.
If we want to get into the details, I am not really sure it just setups the privileged network details, as it also passes info from the virt-handler to the virt-launcher in a odd manner (files).

I find any attempt to make a good name here doomed. But I think the author did not try to give a name that expresses better that this phase1 does in the first place.

> And is so cheap to do it (rename from A to B costs the same independently of what B is) that I don't understand why we wouldn't want to do it.

I guess because not everyone agrees it is a better name that conveys well what this all-in-one function represents.

But if you manage to convince the author, I personally won't object to any naming... all are equally wrong in my view.

**Comment by dankenigsberg** [comment]
> And is so cheap to do it (rename from A to B costs the same independently of what B is) that I don't understand why we wouldn't want to do it.

@maiqueb to me, the cost (and value) of renaming is *not* the same. It takes time and energy to get rid of bad misleading names, and it takes even more to find good ones. In this pr I don't intend to reach nirvana, only to clear some mess.

Once merged, and depending on my capacity, I would like to consider smarter, less trivial changes, and possibly finding better names and better aggregation for the phase functions.

I don't think know if it's polite for me to hold-cancel. I'd appreciate that you or @AlonaKaplan do it to get it in.

**Comment by maiqueb** [comment]
> > And is so cheap to do it (rename from A to B costs the same independently of what B is) that I don't understand why we wouldn't want to do it.
> 
> @maiqueb to me, the cost (and value) of renaming is _not_ the same. It takes time and energy to get rid of bad misleading names, and it takes even more to find good ones. In this pr I don't intend to reach nirvana, only to clear some mess.
> 
> Once merged, and depending on my capacity, I would like to consider smarter, less trivial changes, and possibly finding better names and better aggregation for the phase functions.
> 
> I don't think know if it's polite for me to hold-cancel. I'd appreciate that you or @AlonaKaplan do it to get it in.

You ought to be. Never the less, there it is.

/hold cancel

**Comment by dankenigsberg** [comment]
/retest
(yes, it's lame)

**Review by EdDev — APPROVED** [review]

**Comment by rmohr** [comment]
https://github.com/kubevirt/kubevirt/pull/4838 does not yet resolve all issues and is not yet merged. Merging this PR which clearly suffers from the symptoms there.

**Pull request merged**
